File: backend/app/services/stock_service.py
from sqlalchemy.orm import Session
from fastapi import HTTPException
from ..models.inventory import InventoryItem
from ..models.stock_movement import StockMovement


def adjust_stock(
    db: Session,
    product_id: str,
    quantity: int,
    reason: str,
    user_id: str = "system",
) -> dict:
    item = db.query(InventoryItem).filter(InventoryItem.id == product_id).first()
    if not item:
        raise HTTPException(status_code=404, detail="Producto no encontrado")

    before = item.stock
    new_stock = max(0, before + quantity)
    item.stock = new_stock

    movement = StockMovement(
        product_id=product_id,
        quantity=quantity,
        type="in" if quantity > 0 else "out",
        reason=reason,
        before_stock=before,
        after_stock=new_stock,
        user_id=user_id,
    )
    db.add(movement)
    db.commit()
    db.refresh(item)

    return {"item": item, "movement": movement}


def receive_from_purchase(db: Session, purchase_order: dict, user_id: str) -> dict:
    product_name = purchase_order["item"]
    qty = purchase_order["qty"]
    order_id = purchase_order["id"]
    supplier = purchase_order["supplier"]
    total = purchase_order["total"]

    existing = db.query(InventoryItem).filter(
        InventoryItem.name.ilike(product_name)
    ).first()

    if existing:
        result = adjust_stock(
            db, existing.id, qty,
            f"Recepción OC {order_id} — {supplier}",
            user_id,
        )
        item = result["item"]
    else:
        unit_price = round(total / qty) if qty > 0 else 0
        item = InventoryItem(
            name=product_name,
            sku=f"SKU-{order_id}",
            category="General",
            stock=qty,
            price=unit_price,
            cost=unit_price,
            description=f"Recibido de {supplier} vía {order_id}",
        )
        db.add(item)
        db.flush()

        movement = StockMovement(
            product_id=item.id,
            quantity=qty,
            type="in",
            reason=f"Recepción OC {order_id} — {supplier}",
            before_stock=0,
            after_stock=qty,
            user_id=user_id,
        )
        db.add(movement)
        db.commit()
        db.refresh(item)

    return {"item": item, "order_id": order_id, "qty": qty}


def check_stock_available(db: Session, product_id: str, quantity: int) -> bool:
    item = db.query(InventoryItem).filter(InventoryItem.id == product_id).first()
    if not item:
        return False
    return item.stock >= quantity


def deduct_sale_stock(db: Session, product_id: str, quantity: int, reason: str, user_id: str) -> dict:
    if not check_stock_available(db, product_id, quantity):
        raise HTTPException(
            status_code=400,
            detail="Stock insuficiente para realizar la venta"
        )
    return adjust_stock(db, product_id, -quantity, reason, user_id)

File: dotnet-backend/src/Javaline.Commercial.Infrastructure/Services/StockService.cs
using Javaline.Commercial.Infrastructure.Data;
using Javaline.Commercial.Domain.Entities;
using Microsoft.EntityFrameworkCore;

namespace Javaline.Commercial.Infrastructure.Services;

public interface IStockService
{
    Task<StockAdjustResult> AdjustStockAsync(string productId, int quantity, string reason, string userId);
    Task<List<StockAdjustResult>> ReceiveFromPurchaseAsync(string orderId, string userId);
    Task<bool> CheckStockAvailableAsync(string productId, int quantity);
    Task<StockAdjustResult> DeductSaleStockAsync(string productId, int quantity, string reason, string userId);
}

public class StockAdjustResult
{
    public InventoryItem Item { get; set; } = null!;
    public StockMovement Movement { get; set; } = null!;
}

public class StockService : IStockService
{
    private readonly JavalineDbContext _db;

    public StockService(JavalineDbContext db) => _db = db;

    public async Task<StockAdjustResult> AdjustStockAsync(string productId, int quantity, string reason, string userId)
    {
        var item = await _db.InventoryItems.FindAsync(productId)
            ?? throw new KeyNotFoundException("Producto no encontrado");

        var before = item.Stock;
        var newStock = Math.Max(0, before + quantity);
        item.Stock = newStock;

        var movement = new StockMovement
        {
            Id = Guid.NewGuid().ToString(),
            ProductId = productId,
            Quantity = quantity,
            Type = quantity > 0 ? "in" : "out",
            Reason = reason,
            BeforeStock = before,
            AfterStock = newStock,
            UserId = userId,
            Date = DateTime.UtcNow
        };
        _db.StockMovements.Add(movement);
        await _db.SaveChangesAsync();

        return new StockAdjustResult { Item = item, Movement = movement };
    }

    public async Task<List<StockAdjustResult>> ReceiveFromPurchaseAsync(string orderId, string userId)
    {
        var order = await _db.PurchaseOrders.FindAsync(orderId)
            ?? throw new KeyNotFoundException("Orden no encontrada");

        var results = new List<StockAdjustResult>();

        // Try to find by name, create if not exists
        var existing = await _db.InventoryItems.FirstOrDefaultAsync(i =>
            i.Name.ToLower() == order.Item.ToLower());

        if (existing != null)
        {
            var result = await AdjustStockAsync(existing.Id, (int)order.Qty,
                $"Recepción OC {orderId} — {order.Supplier}", userId);
            results.Add(result);
        }
        else
        {
            var unitPrice = order.Qty > 0 ? Math.Round(order.Total / order.Qty, 2) : 0;
            var item = new InventoryItem
            {
                Id = Guid.NewGuid().ToString(),
                Name = order.Item,
                Sku = $"SKU-{orderId[..Math.Min(8, orderId.Length)]}",
                Category = "General",
                Stock = (int)order.Qty,
                Price = unitPrice,
                Cost = unitPrice,
                Description = $"Recibido de {order.Supplier} vía {orderId}",
                Active = true,
                CreatedAt = DateTime.UtcNow
            };
            _db.InventoryItems.Add(item);

            var movement = new StockMovement
            {
                Id = Guid.NewGuid().ToString(),
                Quantity = (int)order.Qty,
                Type = "in",
                Reason = $"Recepción OC {orderId} — {order.Supplier}",
                BeforeStock = 0,
                AfterStock = (int)order.Qty,
                UserId = userId,
                Date = DateTime.UtcNow
            };
            _db.StockMovements.Add(movement);
            await _db.SaveChangesAsync();
            results.Add(new StockAdjustResult { Item = item, Movement = movement });
        }

        order.Status = "recibido";
        order.ReceivedAt = DateTime.UtcNow;
        await _db.SaveChangesAsync();

        return results;
    }

    public async Task<bool> CheckStockAvailableAsync(string productId, int quantity)
    {
        var item = await _db.InventoryItems.FindAsync(productId);
        return item != null && item.Stock >= quantity;
    }

    public async Task<StockAdjustResult> DeductSaleStockAsync(string productId, int quantity, string reason, string userId)
    {
        if (!await CheckStockAvailableAsync(productId, quantity))
            throw new ApplicationException("Stock insuficiente para realizar la venta");

        return await AdjustStockAsync(productId, -quantity, reason, userId);
    }
}
